docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,287 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
 
 ---
 
-#### 1.1.4 Struct type
+#### 1.1.4 Json type
 
-A special data type is available in Wing called `Struct`. Not to be confused
-with definable structs which are outlined in their own section below. This type
-represents an arbitrary JSON value. This type can currently by assigned to and
-from `any`. During the assignment, the compiler internally knows about the
-length of this type and attempts to read it as either a JSON string or buffer.
+Wing has a data type called `Json`. This type represents an immutable arbitrary [JSON
+value](https://www.json.org/json-en.html) (primitives, arrays and objects).
 
-Since casting is not allowed in Wing, `Struct` data type is `any`'s interface
-with the outside world. `Struct` can be casted back to `any` at any time.  
-All structs can be casted to `Struct` and back to `any` at any time as well.  
-`Struct` cannot be casted to any other type.
+##### 1.1.4.1 Definition
 
-"Struct" is immutable by design. Meaning that once parsed from its `any` buffer,
-it cannot be modified anymore.
+Literals can be defined using the `Json` type initializers:
 
-> ```TS
-> // assuming fetch_some_data_with_jsii() returns "any":
-> let data: Struct = await fetch_some_data_with_jsii();
-> print(data[0]["prop-1"].id);
-> print(data.name);
-> ```
-  
-<details><summary>Equivalent TypeScript Code</summary>
+```js
+let json_string  = Json "hello";
+let json_number  = Json 123;
+let json_bool    = Json true;
+let json_array   = Json [ 1, 2, 3 ];
+let json_obj     = Json { boom: 123 };
+let json_mut_obj = MutJson {
+  hello: 123, 
+  world: [ 1, "cat", 3 ], // <-- heterogenous array
+  boom: { hello: 1233 }
+};
+```
 
-> ```TS
-> const data: any = Object.freeze(await (async () => {
->   try {
->     // deep copy all enumerable properties of the host object
->     return JSON.parse(JSON.stringify(await fetch_some_data_with_jsii()));
->   } catch(err) {
->     throw new Error(`Failed to parse JSON from any: ${err.message}`);
->   }
-> })());
-> console.log(data[0]["prop-1"].id);
-> console.log(data.name);
-> ```
+The type of elements within objects and arrays is still `Json`.
 
-</details>
+To access a field within an object, use the `.` notation:
+
+```js
+let boom: Json = json_obj.boom;
+```
+
+Trying to access a non-existent field will fail at runtime. For example:
+
+```js
+print(json_obj.boom.dude.world);
+// RUNTIME ERROR: Uncaught TypeError: Cannot read properties of undefined (reading 'world')
+```
+
+To access an array element, use the `[]` notation:
+
+```js
+let item2: Json = json_array[2];
+```
+
+It is also possible to assign the native `str`, `num`, `bool` and `Array<T>` values and they will
+implicitly be casted to `Json`:
```

Comment:
No, it is not a valid statement (at least for now). It makes sense to allow implicit assignment from *safe* to *unsafe* but not the other way around.

In this specific example, the compiler can technically tell that the `Json` is just a string and since it is immutable, it can technically allow implicit assignment. But it's not a very interesting case because why would anyone even use this idiom? (they don't need the `Json`).

I'll add this as a clarification to the spec.